### DISCUSSION TITLE
Typeclass Operator Signatures

### DISCUSSION THREAD

**Discussion by ghost** [h3rald/min#102] [Features: Implemented]
My idea:
```
( typeclass natural
	(int :n ==> 'res) ; We already know that it'll return 'bool'.
	(n 0 >= @res)
) ::
```
It would be a super-mixture. This mixture includes inheritance of types or typeclasses. For example, `int` is inherited by `natural` and `natural` becomes more specialized than `int`. Isn't looks good? I think, it would be very good. If we do that, then somethings might be easier than usual.

Interesting Examples:

**Object Oriented Programming(OOP)**:
```
( typeclass object
	(==> 'res)
	(...)
) ::

( typeclass vehicle
	(object :x ==> 'res)
	(...)
) ::

( typeclass car
	(vehicle :x fuel :f rgb :colour ==> 'res)
	(...)
) ::

( symbol efficient
	(car :c bool :detailed-calculating ==> natural :res)
	(...<calculating process>...)
) ::


{"Weaver" :vehicle-name "Fast" :property} ...<vehicle object>...
{"Weaver" :vehicle-name "Fast" :property} 1000 "turqoise" ...<car object>...

{"Weaver" :vehicle-name "Fast" :property} 1000 "turqoise" true efficient puts
; Puts the efficienty score of car.
```

**Recursively Inherited Types/Typeclasses (RIT)** :

```
( typeclass varargs
	(a :lhs varargs :args ==> 'res) ; Means recursive-inheritance
	(true @res)
	
	; It takes two arguments and controls them and goes on with this algorithm.
	
	; Example: 1 2 3 4 5 null -> imagine there are some elements given as 'varargs'
	;  We must control them for 'varargs' constraint:
	
	; 1 2 3 4 5 null -> 1 :lhs <calling the varargs>  :args
	; 2 3 4 5 null -> 2 :lhs <calling the varargs> :args
	; 3 4 5 null -> 3 :lhs <calling the varargs> :args
	; 4 5 null -> 4 :lhs <calling the varargs> :args
	
	; 5 null -> 5 lhs null :args -> Okay! It ended.
	
	
	; It stops when 'args' is 'null'. But it can be specified by anybody.
) ::

; An useful example for it:

( symbol putter
	(string :end-string varargs :args ==>)
	(args puts! end-string puts!)
) ::

"\n"<starter> 1 2 3 4 5 "a" "b" "c" ...<args>... null<delimiter> putter
"" 1 2 + 5 * 6 - putter ; it puts '9' on screen.

; There is no need to use:
"" (1 2 + 5 * 6 -) putter ; Also it doesn't works!
"" (1 2 + 5 * 6 -) -> putter ; This is too ugly!
;And we cannot know about what the last element of applied quot will be returned,
; -->That last element type hidden from constraints(typeclasses).


; There would be a potential to real stack-processing and MORE!.
```

**Comment by h3rald** [comment]
Woah... OK, this is a lot to take in! But definitely very interesting 👍

Here's some initial comments, but I didn't understand all of it...

* I like the idea of using `operator` to define a typeclass! It makes sense as a typeclass is a symbol anyway, so we could create it with `operator`.
* Although we know that a type signature must return a Boolean value, I don't really like the fact that we are omitting the output type. But I understand what you mean. This needs to be improved, but I don't know how... minor issue anyway.
* the biggest issue with all of this is that type classes at the moment MUST take ONLY ONE input value. In your example you are using several... this is a problem because of the way type classes are implemented in Nim, and also because the purpose of a signature is to understand how many input values an operator requires. If a type class could use more than one input value, it would be difficult to understand what goes on the stack...
* I am not sure I understand your example of inheritance. When you are defining a car you are expecting certain values on the stack in a specific order. Why do you consider this OOP? I would have expected something with dictionaries maybe...
* I like the *idea* of recursive type classes, but I don't understand your varargs example. Say you want to accept any number of inputs (that's what varargs is right?), how do you know when to stop popping input values from the stack? That's the problem... do you keep popping items until... you find `null` on the stack?
* Could you write the actual min implementation of the varargs type class??

**Reply by ghost** [reply]
> but I didn't understand all of it

Please say where places didn't you understand? I may explain.

**Reply by ghost** [reply]
> the biggest issue with all of this is that type classes at the moment MUST take ONLY ONE input value. In your example you are using several... this is a problem because of the way type classes are implemented in Nim, and also because the purpose of a signature is to understand how many input values an operator requires. If a type class could use more than one input value, it would be difficult to understand what goes on the stack...

But you looked at the OOP example, right? I'm sorry, but there is no other way(i think) to handle with OOP ,at least. If you know say me how.

> MUST

Why?

**Reply by h3rald** [reply]
> > the biggest issue with all of this is that type classes at the moment MUST take ONLY ONE input value. In your example you are using several... this is a problem because of the way type classes are implemented in Nim, and also because the purpose of a signature is to understand how many input values an operator requires. If a type class could use more than one input value, it would be difficult to understand what goes on the stack...
> 
> 
> 
> But you looked at the OOP example, right? I'm sorry, but there is no other way(i think) to handle with OOP ,at least. If you know say me how.
> 
Well, the main thing is that in your example (if I understand it correctly) you are not using objects, you are using separate values to represent an object right? Why not extend dictionaries instead? They are the closest thing to objects in min...

> 
> > MUST
> 
> Why?

Here's why:

https://github.com/h3rald/min/blob/2ae94bbfa21ede3784aef000155353536dade06b/minpkg/core/utils.nim#L215

The current implementation of type classes relies on that basicValidate method which validates ONE value at a time, within a signature or an `expect` quotation. If you notice, to make it work, I am pushing the value that is being checked on the stack and then pushing the quotation that defines the type class.

The one key things about signatures is that every item should represent a single value, otherwise the programmer doesn't really understand what to put on the stack before an operator...
